File: sdks/node/src/binary.ts
/**
 * Locates the mcparmor binary for the current platform.
 *
 * Search order:
 * 1. MCPARMOR_BIN environment variable (no checksum check — caller's responsibility)
 * 2. Platform-specific optional dependency (SHA256 verified against checksums.ts)
 * 3. Binary on PATH (no checksum check — comes from the user's system)
 */

import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { existsSync, readFileSync } from 'node:fs';
import { BINARY_CHECKSUMS } from './checksums.ts';

/** Maps `${platform}-${arch}` to the corresponding optional npm package name. */
const PLATFORM_PACKAGES: Readonly<Record<string, string>> = {
  'linux-x64':    'mcparmor-linux-x64',
  'linux-arm64':  'mcparmor-linux-arm64',
  'darwin-x64':   'mcparmor-darwin-x64',
  'darwin-arm64': 'mcparmor-darwin-arm64',
  'win32-x64':    'mcparmor-win32-x64',
} as const;

/** Thrown when the mcparmor binary cannot be located by any search method. */
export class BinaryNotFoundError extends Error {
  constructor() {
    super(
      'mcparmor binary not found. Install a platform package (e.g. mcparmor-linux-x64) ' +
      'or set the MCPARMOR_BIN environment variable.',
    );
    this.name = 'BinaryNotFoundError';
  }
}

/** Thrown when the bundled binary fails SHA256 verification. */
export class BinaryChecksumError extends Error {
  constructor(binaryPath: string, expected: string, actual: string) {
    super(
      `mcparmor binary at ${binaryPath} failed SHA256 verification.\n` +
      `  Expected: ${expected}\n` +
      `  Actual  : ${actual}\n` +
      'The binary may have been tampered with. Reinstall the package to fix this.',
    );
    this.name = 'BinaryChecksumError';
  }
}

/**
 * Compute the lowercase hex SHA256 digest of a file.
 *
 * @param filePath - Absolute path to the file to hash.
 * @returns Lowercase hex SHA256 digest string.
 */
function sha256File(filePath: string): string {
  const hash = createHash('sha256');
  hash.update(readFileSync(filePath));
  return hash.digest('hex');
}

/**
 * Verify a binary's SHA256 against the expected checksum for this platform.
 *
 * Skipped when {@link BINARY_CHECKSUMS} is empty (development installs).
 *
 * @param binaryPath - Absolute path to the binary to verify.
 * @throws {BinaryChecksumError} If a checksum is registered for this platform
 *   but the binary's digest does not match.
 */
function verifyChecksum(binaryPath: string): void {
  const platformKey = `${process.platform}-${process.arch}`;
  const expected = BINARY_CHECKSUMS[platformKey];

  // No entry: empty table (dev install) or unrecognised platform — skip both.
  if (expected === undefined) {
    return;
  }

  const actual = sha256File(binaryPath);
  if (actual !== expected) {
    throw new BinaryChecksumError(binaryPath, expected, actual);
  }
}

/**
 * Resolve a path to its absolute, canonical form.
 *
 * On Unix uses `realpath`; falls back to the input string on failure.
 *
 * @param filePath - Path to resolve.
 * @returns Resolved absolute path, or the original path if resolution fails.
 */
function resolvePath(filePath: string): string {
  if (process.platform === 'win32') {
    return filePath;
  }
  try {
    return execFileSync('realpath', [filePath], { encoding: 'utf8' }).trim();
  } catch {
    return filePath;
  }
}

/**
 * Read the binary path from the MCPARMOR_BIN environment variable.
 *
 * @returns The path string, or null if the variable is absent or empty.
 */
function fromEnv(): string | null {
  const value = process.env['MCPARMOR_BIN'];
  return value != null && value.length > 0 ? value : null;
}

/**
 * Resolve the binary bundled inside the platform-specific optional dependency
 * and verify its SHA256 checksum before returning.
 *
 * Uses `require.resolve` so it respects the caller's node_modules layout.
 * Checksum verification is skipped when {@link BINARY_CHECKSUMS} is empty
 * (development installs built from source).
 *
 * @returns Absolute path to the verified binary, or null if the package is not
 *   installed.
 * @throws {BinaryChecksumError} If the binary exists but fails SHA256 verification.
 */
function fromOptionalDependency(): string | null {
  const platformKey = `${process.platform}-${process.arch}`;
  const packageName = PLATFORM_PACKAGES[platformKey];

  if (packageName === undefined) {
    return null;
  }

  let resolved: string;
  try {
    resolved = require.resolve(`${packageName}/bin/mcparmor`);
  } catch {
    return null;
  }

  const absolute = resolvePath(resolved);
  verifyChecksum(absolute);
  return absolute;
}

/**
 * Find the `mcparmor` binary on the system PATH via `which`.
 *
 * @returns Absolute path reported by `which`, or null if not found.
 */
function fromPath(): string | null {
  const whichCommand = process.platform === 'win32' ? 'where' : 'which';

  try {
    const result = execFileSync(whichCommand, ['mcparmor'], { encoding: 'utf8' });
    return result.trim().split('\n')[0]?.trim() ?? null;
  } catch {
    return null;
  }
}

/**
 * Find the mcparmor binary path.
 *
 * Searches in order: MCPARMOR_BIN env var → platform optional dependency → PATH.
 *
 * @returns Absolute path to the mcparmor binary.
 * @throws {BinaryNotFoundError} If the binary cannot be found by any method.
 */
export function findBinary(): string {
  const envPath = fromEnv();
  if (envPath !== null) {
    if (!existsSync(envPath)) {
      throw new BinaryNotFoundError();
    }
    return envPath;
  }

  const depPath = fromOptionalDependency();
  if (depPath !== null) {
    return depPath;
  }

  const pathBinary = fromPath();
  if (pathBinary !== null) {
    return pathBinary;
  }

  throw new BinaryNotFoundError();
}

File: sdks/python/src/mcparmor/_binary.py
"""Binary location resolution for the mcparmor broker."""

import hashlib
import os
import platform
import shutil
import sys
from importlib.resources import files
from pathlib import Path

from ._checksums import BINARY_CHECKSUMS


class BinaryNotFoundError(RuntimeError):
    """Raised when the mcparmor binary cannot be located."""


class BinaryChecksumError(RuntimeError):
    """Raised when the bundled binary fails SHA256 verification."""


_BINARY_NAME = "mcparmor.exe" if sys.platform == "win32" else "mcparmor"
_ENV_VAR = "MCPARMOR_BIN"


def find_binary() -> Path:
    """
    Locate the mcparmor binary using a priority search.

    Search order:
    1. ``MCPARMOR_BIN`` environment variable (no checksum check — caller's responsibility)
    2. Bundled binary in the installed wheel (SHA256 verified against ``_checksums.py``)
    3. Binary on PATH (no checksum check — comes from the user's system)

    Returns:
        Absolute path to the mcparmor binary.

    Raises:
        BinaryNotFoundError: If the binary cannot be found by any method.
        BinaryChecksumError: If the bundled binary fails SHA256 verification.
    """
    candidate = _from_env() or _from_package() or _from_path()
    if candidate is None:
        raise BinaryNotFoundError(
            "mcparmor binary not found. Set MCPARMOR_BIN, install a binary wheel, "
            "or ensure mcparmor is on PATH."
        )
    return candidate


def _from_env() -> Path | None:
    """
    Resolve the binary path from the MCPARMOR_BIN environment variable.

    Returns:
        Absolute path if the env var is set to a non-empty, existing path;
        None otherwise.

    Raises:
        BinaryNotFoundError: If the env var is set but the path does not exist.
    """
    raw = os.environ.get(_ENV_VAR, "").strip()
    if not raw:
        return None

    candidate = Path(raw)
    if not candidate.exists():
        raise BinaryNotFoundError(
            f"{_ENV_VAR} is set to '{raw}' but that path does not exist."
        )
    if not candidate.is_file():
        raise BinaryNotFoundError(
            f"{_ENV_VAR} is set to '{raw}' but that path is not a file."
        )
    if not os.access(candidate, os.X_OK):
        raise BinaryNotFoundError(
            f"{_ENV_VAR} is set to '{raw}' but that file is not executable."
        )
    return candidate.resolve()


def _from_package() -> Path | None:
    """
    Resolve the binary bundled inside the installed wheel and verify its SHA256.

    When ``_checksums.BINARY_CHECKSUMS`` is populated (wheel builds), the
    binary's SHA256 is checked before returning the path. Development installs
    (empty checksums table) skip verification.

    Returns:
        Absolute path to the bundled binary if present, executable, and
        passing the checksum check; None if the binary is not bundled.

    Raises:
        BinaryChecksumError: If the binary exists but fails the SHA256 check.
    """
    try:
        resource = files("mcparmor").joinpath(f"bin/{_BINARY_NAME}")
        candidate = Path(str(resource))
        if not (candidate.exists() and os.access(candidate, os.X_OK)):
            return None
        candidate = candidate.resolve()
        _verify_checksum(candidate)
        return candidate
    except (TypeError, ModuleNotFoundError):
        return None


def _verify_checksum(binary_path: Path) -> None:
    """
    Verify a binary's SHA256 against the expected checksum for this platform.

    Skipped when ``BINARY_CHECKSUMS`` is empty (development installs).

    Args:
        binary_path: Absolute path to the binary to verify.

    Raises:
        BinaryChecksumError: If a checksum is registered for this platform
            but the binary's digest does not match.
    """
    if not BINARY_CHECKSUMS:
        return

    platform_tag = _current_platform_tag()
    expected = BINARY_CHECKSUMS.get(platform_tag)
    if expected is None:
        # No checksum registered for this platform — skip verification.
        return

    actual = _sha256(binary_path)
    if actual != expected:
        raise BinaryChecksumError(
            f"mcparmor binary at {binary_path} failed SHA256 verification.\n"
            f"  Expected: {expected}\n"
            f"  Actual  : {actual}\n"
            "The binary may have been tampered with. Reinstall the package to fix this."
        )


def _sha256(path: Path) -> str:
    """Return the lowercase hex SHA256 digest of a file."""
    digest = hashlib.sha256()
    with path.open("rb") as fh:
        for chunk in iter(lambda: fh.read(65536), b""):
            digest.update(chunk)
    return digest.hexdigest()


def _current_platform_tag() -> str:
    """
    Return a platform tag matching the wheel naming convention.

    Used to look up the expected checksum in ``BINARY_CHECKSUMS``.
    """
    system = platform.system().lower()
    machine = platform.machine().lower()

    # Normalise machine names to wheel convention.
    machine_map = {
        "x86_64": "x86_64",
        "amd64": "x86_64",
        "aarch64": "aarch64",
        "arm64": "arm64",  # macOS convention
    }
    machine = machine_map.get(machine, machine)

    if system == "darwin":
        major, minor = platform.mac_ver()[0].split(".")[:2]
        return f"macosx_{major}_{minor}_{machine}"
    if system == "linux":
        return f"manylinux_2_28_{machine}"
    if system == "windows":
        return f"win_{machine}"
    return f"{system}_{machine}"


def _from_path() -> Path | None:
    """
    Resolve the binary from the system PATH.

    Returns:
        Absolute path to the binary if found on PATH; None otherwise.
    """
    found = shutil.which("mcparmor")
    if found is None:
        return None
    return Path(found).resolve()
